### PR TITLE
[WIP] enable temp sensor

### PR DESCRIPTION
**Pull request by sumnerboy12**

### CODE REVIEW

**Comment by sumnerboy12** [comment]
@moinmoin-sh - I have been working on the new `Rack32` library in this branch and have pushed it here for you to have a look at. I will pull the `OXRS_Rack32` files out and into a standalone library before merging this, but thought it was probably easier for you to test if it was all included in this repo directly.

The general idea is that all the `.setMqttXXX()` stuff will be replaced by the Admin UI - which will likely write the MQTT config to persistent flash during on-boarding and thus all you need to add to your firmware code is the `rack32.begin()`, setup the display ports and then `rack32.loop()` (along with the display port updates).

Let me know your thoughts.

File: OXRS-BMD-SmokeDetector-ESP32-FW.ino [hunk 1]
```diff
@@ -457,9 +367,9 @@ void publishEvent(uint8_t index, char * type, char * event)
   uint8_t channel = getChannel(index);
 
   // Show event on screen
-  char display[32];
-  sprintf_P(display, PSTR("idx:%2d %s %s   "), index, type, event);
-  screen.show_event(display);
+//  char display[32];
```
Comment (sumnerboy12) [inline OXRS-BMD-SmokeDetector-ESP32-FW.ino:357]: FYI - I have moved this inside the `Rack32` library on my local machine (where I have moved `Rack32` into a stand alone library). The `Rack32` library checks if the JSON payload has `index`, `type` and `event` fields and if so generates an event for displaying on the LCD.

**Comment by moinmoin-sh** [comment]
It compiles, but doesn't run on my HW. It hangs in an endless boot loop.

I commented out 
```
//  rack32.setMqttBroker(MQTT_BROKER, MQTT_PORT);
//  rack32.setMqttAuth(MQTT_USERNAME, MQTT_PASSWORD);
// rack32.setMqttTopicPrefix(MQTT_TOPIC_PREFIX);
// rack32.setMqttTopicSuffix(MQTT_TOPIC_SUFFIX);
```
and then it boots, shows the screen but then of course doesn't get a mqtt connection.

I believe this is no good
```
// Rack32 handler
OXRS_Rack32 rack32(FW_NAME, FW_SHORT_NAME, FW_MAKER_CODE, FW_VERSION, FW_CODE);

```

because it translates to 
`OXRS_Rack32 rack32("OXRS-BMD-SmokeDetector-ESP32-FW", "Smoke Detector",  ....  );`


which I think creates the strings on the stack. When you store these addresses in the contstuctor of the class you store stack addresses which are reused .

Not sure ...
What do you think?  Wonder why it worked on your site. Did I do something wrong or did I miss something?

**Comment by moinmoin-sh** [comment]
The above doesn't seem to be true.
nailed it down to commenting out these 
```
//  rack32.setMqttAuth(MQTT_USERNAME, MQTT_PASSWORD);
// rack32.setMqttTopicPrefix(MQTT_TOPIC_PREFIX);
// rack32.setMqttTopicSuffix(MQTT_TOPIC_SUFFIX);
```
maybe the NULL ptr is not good

Do you have the MQTT DEFINES set to something different than NULL?

In the "old" version you do the settings only after checking the DEFINEs for != NULL

 ```
 // Set credentials and any extra topic parts specified
  if (MQTT_USERNAME     != NULL) { mqtt.setAuth(MQTT_USERNAME, MQTT_PASSWORD); }
  if (MQTT_TOPIC_PREFIX != NULL) { mqtt.setTopicPrefix(MQTT_TOPIC_PREFIX); }
  if (MQTT_TOPIC_SUFFIX != NULL) { mqtt.setTopicSuffix(MQTT_TOPIC_SUFFIX); }

```

**Comment by sumnerboy12** [comment]
Yeah sorry about that, should have mentioned it. I haven't added the NULL
handling like I had in the firmware when setting those config fields.

Wanted to move the checks inside the library so each piece of firmware
didn't have to do the same checks.

Apologies!

Does it work otherwise? What do you think of the code?

**Comment by moinmoin-sh** [comment]
OK, I passed the test.

I think it's a good idea to move the "every time needed" code into a lib. So it can be reused, doesn't need to be written repeatedly and enhancements/fixes can be deployed easy. On the other hand it might not be easy for newcomers to understand the system and how things work together, The sketch looks different than usual. May be we need a diagram that shows the different layers and how they comunicate.

I had only a short time for testing, will do more tomorrow as time permits. I missed of course the event display, but you mentioned that already.

I think the `config.h.example` needs some cleanup to make sure nothing is double defined.

**Comment by sumnerboy12** [comment]
Yep it is a fine line between shielding the user from complications, and making it so simple they don't have any idea what is going on. To be honest, the main motivation for doing this is selfish!! I was tired of making a change to one of the firmware versions and having to replicate it across the other 3/4 firmware we have.

`config.h.example` hasn't been updated yet - I will fix that up when I move this to a dedicated library and update all the firmware to point to it properly.

I have asked Jon to create the new repo so just waiting on that.

**Comment by sumnerboy12** [comment]
This was superseded by the move to the new Rack32 library.

**Pull request closed without merging**